packages/block-library/src/term-template/edit.js

```diff
@@ -0,0 +1,277 @@
+			style={ style }
+		>
+			{ termName }
+		</li>
+	);
+}
+
+const MemoizedTermTemplateBlockPreview = memo( TermTemplateBlockPreview );
+
+/**
+ * Builds a hierarchical tree structure from flat terms array.
+ *
+ * @param {Array} terms Array of term objects.
+ * @return {Array} Tree structure with parent/child relationships.
+ */
+function buildTermsTree( terms ) {
+	const termsById = {};
+	const rootTerms = [];
+
+	terms.forEach( ( term ) => {
+		termsById[ term.id ] = {
+			term,
+			children: [],
+		};
+	} );
+
+	terms.forEach( ( term ) => {
+		if ( term.parent && termsById[ term.parent ] ) {
+			termsById[ term.parent ].children.push( termsById[ term.id ] );
+		} else {
+			rootTerms.push( termsById[ term.id ] );
+		}
+	} );
+
+	return rootTerms;
+}
+
+/**
+ * Renders a single term node and its children recursively.
+ *
+ * @param {Object}   termNode   Term node with term object and children.
+ * @param {Function} renderTerm Function to render a single term.
+ * @return {JSX.Element} Rendered term node.
+ */
+function renderTermNode( termNode, renderTerm ) {
+	const children =
+		termNode.children.length > 0 ? (
+			<ul>
+				{ termNode.children.map( ( childNode ) =>
+					renderTermNode( childNode, renderTerm )
+				) }
+			</ul>
+		) : null;
+
+	return (
+		<>
+			{ renderTerm( termNode.term ) }
+			{ children }
+		</>
+	);
+}
+
+export default function TermTemplateEdit( {
+	clientId,
+	context: {
+		termQuery: {
+			taxonomy,
+			order,
+			orderBy,
+			hideEmpty,
+			hierarchical,
+			parent,
+		} = {},
+	},
+} ) {
+	const [ activeBlockContextId, setActiveBlockContextId ] = useState();
+
+	const queryArgs = {
+		order,
+		orderby: orderBy,
+		hide_empty: hideEmpty,
+	};
+
+	const { records: terms, isResolving } = useEntityRecords(
+		'taxonomy',
+		taxonomy,
+		queryArgs
+	);
+
+	// Filter to show only top-level terms if "Show only top-level terms" is enabled.
+	const filteredTerms = useMemo( () => {
+		if ( ! terms || parent !== 0 ) {
+			return terms;
+		}
+		return terms.filter( ( term ) => ! term.parent );
+	}, [ terms, parent ] );
+
+	const { blocks } = useSelect(
+		( select ) => ( {
+			blocks: select( blockEditorStore ).getBlocks( clientId ),
+		} ),
+		[ clientId ]
+	);
+
+	const blockContexts = useMemo(
+		() =>
+			filteredTerms?.map( ( term ) => ( {
+				taxonomy,
+				termId: term.id,
+				classList: `term-${ term.id }`,
+			} ) ),
+		[ filteredTerms, taxonomy ]
+	);
+
+	const blockProps = useBlockProps( {
+		className: clsx( 'wp-block-term-template' ),
+	} );
+
+	if ( isResolving ) {
+		return (
+			<p { ...blockProps }>
+				<Spinner />
+			</p>
+		);
+	}
+
+	if ( ! filteredTerms?.length ) {
+		return <p { ...blockProps }> { __( 'No terms found.' ) }</p>;
+	}
+
+	const renderTerm = ( term ) => {
+		const blockContext = {
+			taxonomy,
+			termId: term.id,
+			classList: `term-${ term.id }`,
+		};
+
+		return (
+			<BlockContextProvider key={ term.id } value={ blockContext }>
+				{ term.id ===
```

Comment:
It might be worth extracting this condition to a function, so that it can have a name the explains more clearly what it's doing.

Reply:
Thanks! Created a new function for this in https://github.com/WordPress/gutenberg/pull/70747/commits/7d9b497cd275de1f1f6f94c14630ab8e1c5da543.